Xamarin.Forms.Platform.Android/FastRenderers/MediaElementRenderer.cs

```diff
@@ -0,0 +1,481 @@
+		{
+			if (_defaultLabelFor is null)
+			{
+				_defaultLabelFor = LabelFor;
+			}
+
+			LabelFor = (int)(id ?? _defaultLabelFor);
+		}
+
+		void SetTracker(VisualElementTracker tracker)
+		{
+			_tracker = tracker;
+		}
+
+		void UpdateBackgroundColor()
+		{
+			SetBackgroundColor(Element.BackgroundColor.ToAndroid());
+		}
+
+		void IVisualElementRenderer.UpdateLayout() => _tracker?.UpdateLayout();
+
+		protected override void Dispose(bool disposing)
+		{
+			if (_isDisposed)
+			{
+				return;
+			}
+
+			_isDisposed = true;
+
+			ReleaseControl();
+
+			if (disposing)
+			{
+				SetOnClickListener(null);
+				SetOnTouchListener(null);
+
+				_automationPropertiesProvider?.Dispose();
+				_tracker?.Dispose();
+
+				if (Element != null)
+				{
+					Element.PropertyChanged -= OnElementPropertyChanged;
+
+					if (Platform.GetRenderer(Element) == this)
+						Element.ClearValue(Platform.RendererProperty);
+				}
+			}
+
+			base.Dispose(disposing);
+		}
+
+		void OnElementChanged(ElementChangedEventArgs<MediaElement> e)
+		{
+			if (e.OldElement != null)
+			{
+				
+			}
+
+			if (e.NewElement != null)
+			{
+				this.EnsureId();
+
+				UpdateKeepScreenOn();
+				UpdateLayoutParameters();
+				UpdateShowPlaybackControls();
+				UpdateSource();
+				UpdateBackgroundColor();
+
+				ElevationHelper.SetElevation(this, e.NewElement);
+			}
+
+			ElementChanged?.Invoke(this, new VisualElementChangedEventArgs(e.OldElement, e.NewElement));
+		}
+
+		void MetadataRetrieved(object sender, EventArgs e)
+		{
+			Controller.Duration = _view.DurationTimeSpan;
+			Controller.VideoHeight = _view.VideoHeight;
+			Controller.VideoWidth = _view.VideoWidth;
+
+			Device.BeginInvokeOnMainThread(UpdateLayoutParameters);
+		}
+
+		void OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
+		{
+			switch(e.PropertyName)
+			{
+				case nameof(MediaElement.Aspect):
+					UpdateLayoutParameters();
+					break;
+
+				case nameof(MediaElement.IsLooping):
+					if (_mediaPlayer != null)
+					{
+						_mediaPlayer.Looping = MediaElement.IsLooping;
+					}
+					break;
+
+				case nameof(MediaElement.KeepScreenOn):
+					UpdateKeepScreenOn();
+					break;
+
+				case nameof(MediaElement.ShowsPlaybackControls):
+					UpdateShowPlaybackControls();
+					break;
+
+				case nameof(MediaElement.Source):
+					UpdateSource();
+					break;
+			}
+
+			ElementPropertyChanged?.Invoke(this, e);
+		}
+
+		public void RegisterEffect(Effect effect)
+		{
+			_effectControlProvider.RegisterEffect(effect);
+		}
+
+		void UpdateKeepScreenOn()
+		{
+			_view.KeepScreenOn = MediaElement.KeepScreenOn;
+		}
+
+		void UpdateShowPlaybackControls()
+		{
+			_controller.Visibility = MediaElement.ShowsPlaybackControls ? ViewStates.Visible : ViewStates.Gone;
+		}
+
+		void UpdateSource()
+		{
+			if (MediaElement.Source != null)
+			{
+				if (MediaElement.Source is UriMediaSource uriSource)
+				{
+					if (uriSource.Uri.Scheme == "ms-appx")
```

Comment:
I am worried that this may cause confusion. This is very UWP-ish and appx has no meaning outside UWP - and the ms- prefix.

Reply:
The original code was built to only use Uri for source, so I replicated the custom ms- schemes across the other platforms. When I reworked it to use typed *MediaSource classes I left it in, but I can see the confusion. It was useful because it allowed a common method to reference files in the app package itself, along with common folders like temp and local data without knowing absolute paths.